lib/rewards/server/subscribers/rewards-subscriber.js

```diff
@@ -0,0 +1,190 @@
+const { sc, Logger } = require('@reldens/utils');
+const { Modifier } = require('@reldens/modifiers');
+const { Reward } = require('../models/reward');
+
+class RewardsSubscriber
+{
+    static async giveRewards(playerSchema, targetObject, eventsManager)
+    {
+        let rewards = sc.get(targetObject, 'rewards', false);
+
+        let winningRewards = this.getWinningRewards(rewards);
+        if (0 === winningRewards.length) {
+            return false;
+        }
+        let repositories = this.getRepositories(targetObject);
+
+        let itemRewards = [];
+        for (let winningReward of winningRewards) {
+            let giveRewardResults = await this.giveReward(winningReward, repositories, playerSchema, async (item) => {
+                itemRewards.push({
+                    item,
+                    reward: winningReward
+                });
+            });
+            if (0 < giveRewardResults.length) {
+                Logger.error('There were error(s) on the reward ' + winningReward.id +
+                    ' given to the player ' + playerSchema.id, giveRewardResults);
+            }
+        }
+
+        await this.emitEvent(eventsManager, itemRewards, playerSchema, targetObject);
+        await this.updateUniqueRewards(winningRewards, targetObject.dataServer);
+        return true;
+    }
+
+    static async emitEvent(eventsManager, itemRewards, playerSchema, targetObject)
+    {
+        if (0 === itemRewards.length) {
```

Comment:
I think you already removed this one, but just to note it keep events in the same function that happens